### DISCUSSION TITLE
How to sample the palette of an image

### DISCUSSION THREAD

**Discussion by mmower** [libvips/libvips#4016] [General]
For an art project I want to sample the colours in an image, to get the 26 most used colours.

The naive approach I hit on is just getting individual pixels and counting colours.

But the getPixel API seems appreciably slow, even resizing to 100x100 i'm not sure it's practical for the cheapish hardware option I have in mind. (I was testing it on an M2 Max Mac Studio with 64GB RAM if relevant).

Is there a better approach or a faster API?

For ref: I just came across this library today and am not very experienced in image management.

Thanks

Matt

**Comment by mmower** [comment]
Okay I missed some details. I am actually using the Vix Elixir library that wraps libvips.

Sampling 100x100 takes about 1.5s on this machine.

IEx session with the test code:

[testing_vixvips.exs.zip](https://github.com/user-attachments/files/15978580/testing_vixvips.exs.zip)

**Comment by jcupitt** [comment]
Hi @mmower,

Yes, `getpoint` is just for emergencies, it's too slow (for technical reasons) to use in a loop over an image. 

I would find the 3d histogram, then get the 26 highest values. In python (I don't know vix well) it'd be:

```python
#!/usr/bin/env python3

import sys
import pyvips

# number of bins in each axis 
BINS = 16
BIN_SIZE = 256 / BINS

image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")
hist = image.hist_find_ndim(bins=BINS).bandunfold().copy_memory()
result = hist.max(size=26, x_array=True, y_array=True)

print(f"26 most common colours:")
for x, y in zip(result[1]["x_array"], result[1]["y_array"]):
    r = BIN_SIZE / 2 + BIN_SIZE * (x // BINS)
    b = BIN_SIZE / 2 + BIN_SIZE * (x % BINS)
    g = BIN_SIZE / 2 + BIN_SIZE * y
    count = hist(x, y)[0]

    print(f"count = {count}, rgb = {r}, {g}, {b}")
```

I see:

```
$ time ./most-used.py ~/pics/st-francis.jpg
26 most common colours:
count = 8344730.0, rgb = 88.0, 72.0, 56.0
count = 8362161.0, rgb = 88.0, 56.0, 24.0
count = 8794452.0, rgb = 104.0, 104.0, 72.0
count = 8929635.0, rgb = 24.0, 8.0, 8.0
count = 8930301.0, rgb = 88.0, 88.0, 56.0
count = 9117394.0, rgb = 72.0, 72.0, 40.0
count = 10409121.0, rgb = 152.0, 152.0, 120.0
count = 11851995.0, rgb = 120.0, 120.0, 72.0
count = 12207485.0, rgb = 136.0, 120.0, 72.0
count = 12300594.0, rgb = 120.0, 120.0, 88.0
count = 12333095.0, rgb = 104.0, 88.0, 56.0
count = 13234239.0, rgb = 40.0, 24.0, 8.0
count = 13285376.0, rgb = 104.0, 88.0, 40.0
count = 13674498.0, rgb = 136.0, 136.0, 88.0
count = 15019726.0, rgb = 120.0, 104.0, 56.0
count = 16018057.0, rgb = 136.0, 136.0, 104.0
count = 16079152.0, rgb = 8.0, 8.0, 8.0
count = 17166827.0, rgb = 56.0, 56.0, 40.0
count = 18143106.0, rgb = 72.0, 56.0, 40.0
count = 18246135.0, rgb = 72.0, 56.0, 24.0
count = 18584306.0, rgb = 88.0, 72.0, 40.0
count = 19437765.0, rgb = 40.0, 24.0, 24.0
count = 29976858.0, rgb = 56.0, 40.0, 24.0
count = 32555817.0, rgb = 24.0, 24.0, 24.0
count = 35259528.0, rgb = 24.0, 24.0, 8.0
count = 35932160.0, rgb = 40.0, 40.0, 24.0

real	0m7.237s
user	0m20.895s
sys	0m0.659s
```

So 7s for a 30,000 x 30,000 pixel jpeg.

It works by making a 16 x 16 x 16 element matrix of buckets, then scanning the image and for each RGB pixel, incrementing the element in the corresponding bucket. Then the coordinates of the 26 highest histogram values are the RGBs of the 26 most common colours. You don't get huge accuracy, but it's very quick.

You could turn the number of buckets up, perhaps to 64. You could give the image a little blur beforehand to get noise down. You could make the histogram in a more perceptually linear colourspace to get more uniform discrimination.

**Reply by jcupitt** [reply]
The other obvious algorithm would be some variation on k-means clustering, but that's not built into libvips, unfortunately.

https://en.wikipedia.org/wiki/K-means_clustering

**Reply by mmower** [reply]
Thanks, John. You're a star. I had read something about this in another thread but couldn't quite grok it. You've given me a great jumping off point and I can read Python well enough to translate it. For this project I suspect this approach will be good enough and cheap enough.

**Reply by mmower** [reply]
Here's my WIP translation of your Python example and I am a little stuck about where the `max` method is coming from in `hist.max(…)`.

```
{:ok, image} = Vix.Vips.Image.new_from_file(image_file1)

bins = 16
bin_size = round(256/bins)

{:ok, hist} = Vix.Vips.Operation.hist_find_ndim(image, bins: bin_size)
{:ok, bands} = Vix.Vips.Operation.bandunfold(hist)
{:ok, mem_img} = Vix.Vips.Image.copy_memory(bands)
# something here that represents hist.max
```

**Reply by jcupitt** [reply]
You'd need to ask the vix maintainers, I don't know how they implemented multiple returns from operations.

**Reply by jcupitt** [reply]
Ah you can make it a little simpler:

```python
#!/usr/bin/env python3

import sys
import pyvips

# number of bins in each axis 
BINS = 16
BIN_SIZE = 256 / BINS
  
image = pyvips.Image.new_from_file(sys.argv[1], access="sequential")
hist = image.hist_find_ndim(bins=BINS).bandunfold()
_, result = hist.max(size=26, out_array=True, x_array=True, y_array=True)
    
print(f"26 most common colours:")
for v, x, y in zip(result["out_array"], result["x_array"], result["y_array"]):
    r = BIN_SIZE / 2 + BIN_SIZE * (x // BINS)
    b = BIN_SIZE / 2 + BIN_SIZE * (x % BINS)
    g = BIN_SIZE / 2 + BIN_SIZE * y
    
    print(f"count = {v}, rgb = {r}, {g}, {b}")
```

**Reply by mmower** [reply]
Hey, John. I'm not sure if it was you but someone mentioned that this approach in SRGB space might result in "dimmer" palettes and that I should try LAB. My experiment with that lead to "a suffusion of brown" and it was suggested this might relate to SRGB being 0..255 while LAB is 1…100, -128…128. My naive approach was to convert to LAB before giving it to this code and then converting the results back to SRGB. Should that even work?

**Reply by jcupitt** [reply]
You'd need to:

- turn the image to lab
- stretch and squash the ranges (L is 0-100, a* is -128 to +128, b* the same) to fir into three 8-bit channels (all 0-255) 
- a* and b* are only theoretically that large, probably you'd want just -50 to +50, something like that
- find the histogram and then the maxima
- reverse the stretch and squash to recover lab
- turn lab back to rgb

It'd be better, but probably only marginally. 

I'd look at k-means clustering next. A very simple way to try this is with `gifsave`, for example:

```
$ vips copy k2.jpg x.gif[bitdepth=5]
$ vipsheader -f gif-palette x.gif
-12308689 -10255192 -13296356 -11578544 -13352381 -12176310 -11176299 -11782088 -15923190 -15459549 -14537425 -8019262 -9137229 -459786 -1118232 -12955026 -11902103 -5784635 -11770742 -3742995 -5255716 -13418907 -10326163 -5985104 -1711658 -14867122 -9928313 -2961979 -6317414 -7243645 -7304567 -8757399 
```

That's saving with 5 bits per pixel, so 2 ^ 5, or 32 colours. The palette displays the 32 colours it chose as signed 32-bit ints -- turn those to unsigned hex to get the RGB colours, so `-12308689` is `#BBD0D1`. You could just pick 26 of those 32 randomly.

**Reply by mmower** [reply]
Thanks, John.
